_posts/2020-01-27-serverless-express-openapi.md

```diff
@@ -0,0 +1,522 @@
+        Variables:
+          BUCKET: !Ref CommentsBucket
+      Policies:
+        - S3CrudPolicy:
+            BucketName: !Ref CommentsBucket
+      Events:
+        Get:
+          Type: HttpApi # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
+          Properties:
+            Path: /
+            Method: get
+            ApiId: !Ref GuestBookApi
+        Post:
+          Type: HttpApi
+          Properties:
+            Path: /
+            Method: post
+            ApiId: !Ref GuestBookApi
+
+Outputs:
+  GuestBookApi:
+    Description: API Gateway endpoint URL for Prod stage for Hello World function
+    Value:
+      Fn::Sub: https://${GuestBookApi}.execute-api.${AWS::Region}.amazonaws.com/
+  ExpressBackendFunction:
+    Description: Express Backend Lambda Function ARN
+    Value: !Sub ExpressBackend.Arn
+  ExpressBackendIamRole:
+    Description: Implicit IAM Role created for Hello World function
+    Value: !Sub ExpressBackendFunctionRole.Arn
+```
+
+We actually didn't need to change all that much.
+
+We needed to explicitly add a new resource, `AWS::Serverless::HttpApi`. I say 'explicitly', because SAM actually creates one implicitly with the logical ID 'ServerlessHttpApi', if there isn't one specified in the template. If we want to use an OpenAPI document though, we need to specify one.
+
+We use a common trick here to import our separately defined specification using the 'Fn::Transform' macro. Specifying this macro calls CloudFormation to fetch the file specified in the 'Location' key at execution time, and this is typically an S3 Bucket location. The SAM deploy process recognises that we have specified a local file will upload the file to an S3 bucket and rewrite the reference at deploy time.
+
+Performing `sam deploy && sam build` should work successfully with no issues.
+
+# Generating Client Code
+
+Our next step is to generate client code from the specification. There are a lot tools out there to do this, but we will use the [OpenAPI generator](https://github.com/OpenAPITools/openapi-generator). Installation instructions are available [here](https://openapi-generator.tech/docs/installation.html).
+
+We are going to take a bit of detour from javascript and use the typescript-axios generator as it seems to produce a slightly better client. Let's take it for a spin!
+
+```bash
+openapi-generator generate \
+  -i api.yaml                         `# specify location of OpenAPI document` \
+  -g typescript-axios                 `# specify which code generator to use` \
+  -o ./client-ts                      `# output directory` \
+  -p prependFormOrBodyParameters=true `# generator specific parameters` \
+  -p supportsES6=true \
+  -p npmName=guestbook-client \
+  -p npmVersion=1.0.0
+```
+
+Assuming this worked, it should have output a bunch of files into the 'client-ts' directory. Admittedly, it looked a little a messy, so I made a couple of changes.
```

Comment:
Assuming? I'd probably adjust the sentence slightly to say:

Running the above generate command outputs a bunch of files into the 'client-ts' directory. It looks a little a messy right now, so let's make a couple of changes.

Reply:
I adjusted the sentence to be in present tense since the user is following a tutorial (in case you're wondering)

Reply:
Some of the code-generators can be a little flaky. The Azure one is the best but it's a pain in the ass to install properly. There's a bit of 'fingers crossed' beyond this point.